### ISSUE TITLE
Need to support read filters for nested lists with no keys

### ISSUE THREAD

**Issue by abhikeshav** [CiscoDevNet/ydk-gen#399]
Currently, YDK does not creating a read filter for payloads like the below, where `vrf, af, network` are all lists:
```
<filter>
<bgp xmlns="http://cisco.com/ns/yang/Cisco-IOS-XR-ipv4-bgp-oper">
                        <instances>
                            <instance>
                                <instance-active>
                                    <vrfs>
                                        <vrf>
                                            <afs>
                                              <af>
                                                <networks>
                                                  <network>
                                                    <network/>
                                                  </network>
                                                </networks>
                                              </af>
                                            </afs>
                                        </vrf>
                                    </vrfs>
                                </instance-active>
                            </instance>
                        </instances>
                    </bgp>
</filter>
```
Adding support for the above will be useful for read operations. Currently only leaving out the key for the most deeply nested class is supported for `crud.read`:

```python
from ydk.models.cisco_ios_xr import Cisco_IOS_XR_ip_rib_ipv4_oper
from ydk.providers import NetconfServiceProvider
 
rib = Cisco_IOS_XR_ip_rib_ipv4_oper.Rib()
 
vrf = Cisco_IOS_XR_ip_rib_ipv4_oper.Rib.Vrfs.Vrf()
vrf.vrf_name='default'
 
af = Cisco_IOS_XR_ip_rib_ipv4_oper.Rib.Vrfs.Vrf.Afs.Af()
af.af_name = 'IPv4'
 
saf = Cisco_IOS_XR_ip_rib_ipv4_oper.Rib.Vrfs.Vrf.Afs.Af.Safs.Saf()
saf.saf_name='Unicast'
 
table_name = Cisco_IOS_XR_ip_rib_ipv4_oper.Rib.Vrfs.Vrf.Afs.Af.Safs.Saf.IpRibRouteTableNames.IpRibRouteTableName()
table_name.route_table_name = 'default'
 
route = Cisco_IOS_XR_ip_rib_ipv4_oper.Rib.Vrfs.Vrf.Afs.Af.Safs.Saf.IpRibRouteTableNames.IpRibRouteTableName.Routes.Route()
#route.address='0.0.0.0' # Do not set the route address. This works with crud.read currently
 
table_name.routes.route.append(route)
saf.ip_rib_route_table_names.ip_rib_route_table_name.append(table_name)
af.safs.saf.append(saf)
vrf.afs.af.append(af)
rib.vrfs.vrf.append(vrf)
```

Something like the below should also work with `crud.read`:
```python
from ydk.models.cisco_ios_xr import Cisco_IOS_XR_ip_rib_ipv4_oper
from ydk.providers import NetconfServiceProvider
 
rib = Cisco_IOS_XR_ip_rib_ipv4_oper.Rib()
 
vrf = Cisco_IOS_XR_ip_rib_ipv4_oper.Rib.Vrfs.Vrf()
 
af = Cisco_IOS_XR_ip_rib_ipv4_oper.Rib.Vrfs.Vrf.Afs.Af()
 
saf = Cisco_IOS_XR_ip_rib_ipv4_oper.Rib.Vrfs.Vrf.Afs.Af.Safs.Saf()
 
table_name = Cisco_IOS_XR_ip_rib_ipv4_oper.Rib.Vrfs.Vrf.Afs.Af.Safs.Saf.IpRibRouteTableNames.IpRibRouteTableName()
 
route = Cisco_IOS_XR_ip_rib_ipv4_oper.Rib.Vrfs.Vrf.Afs.Af.Safs.Saf.IpRibRouteTableNames.IpRibRouteTableName.Routes.Route()
 
table_name.routes.route.append(route)
saf.ip_rib_route_table_names.ip_rib_route_table_name.append(table_name)
af.safs.saf.append(saf)
vrf.afs.af.append(af)
rib.vrfs.vrf.append(vrf)
```

**Comment by abhikeshav** [comment]
This works now with `YFilter` in `0.6.0`. See http://ydk.cisco.com/py/docs/guides/netconf_operations.html for example

**Comment by thenetworkingguy** [comment]
Hello,

Can you please give an example on the YFilter.Read operation ?

Apologies for not opening another post but is it possible for you to check the below code, for the love of GOD, i can't figure out the simple task of getting an ISIS neighbor table 

```
    from ydk.services import CodecService
    from ydk.providers import CodecServiceProvider
    from ydk.models.cisco_ios_xr import Cisco_IOS_XR_clns_isis_oper

    codec_provider = CodecServiceProvider(type="xml")
    codec = CodecService()
    isis = Cisco_IOS_XR_clns_isis_oper.Isis()

    c_instance = Cisco_IOS_XR_clns_isis_oper.Isis.Instances.Instance()
    c_instance.instance_name = "1"

    c_neighbor = Cisco_IOS_XR_clns_isis_oper.Isis.Instances.Instance.Neighbors.Neighbor()

    c_instance.neighbors.neighbor.append(c_neighbor)
    isis.instances.instance.append(c_instance)
    print(codec.encode(codec_provider, c_neighbor))
```

I get the following error on running the code, I am guessing its complaining that the inner most key is not set but i did set the instance key

```
Traceback (most recent call last):
  File "lpts.py", line 134, in <module>
    print(codec.encode(codec_provider, c_neighbor))
  File "/home/ubuntu/.virtualenvs/ydk-py/local/lib/python2.7/site-packages/ydk/errors/error_handler.py", line 112, in helper
    return func(self, provider, entity, *args, **kwargs)
  File "/home/ubuntu/.virtualenvs/ydk-py/local/lib/python2.7/site-packages/ydk/services/codec_service.py", line 74, in encode
    return self._encode(provider, entity_holder, pretty, subtree)
  File "/home/ubuntu/.virtualenvs/ydk-py/local/lib/python2.7/site-packages/ydk/services/codec_service.py", line 101, in _encode
    data_node = _get_data_node_from_entity(entity, root_schema)
  File "/home/ubuntu/.virtualenvs/ydk-py/local/lib/python2.7/site-packages/ydk/models/cisco_ios_xr/Cisco_IOS_XR_clns_isis_oper.py", line 6841, in get_entity_path
    raise YPYModelError("ancestor cannot be None as one of the ancestors is a list")
ydk.errors.YPYModelError: ancestor cannot be None as one of the ancestors is a list
```

**Comment by abhikeshav** [comment]
Can you try the below for example of `YFilter.read`? It can be used with `subtree` option.

The issue with your code was that codec service only accepts top level containers (`isis` instead of `c_neighbor`:

```python
from ydk.services import CodecService
from ydk.providers import CodecServiceProvider
from ydk.models.cisco_ios_xr import Cisco_IOS_XR_clns_isis_oper
from ydk.filters import YFilter

codec_provider = CodecServiceProvider(type="xml")
codec = CodecService()
isis = Cisco_IOS_XR_clns_isis_oper.Isis()

c_instance = Cisco_IOS_XR_clns_isis_oper.Isis.Instances.Instance()
c_instance.instance_name = "1"

c_neighbor = Cisco_IOS_XR_clns_isis_oper.Isis.Instances.Instance.Neighbors.Neighbor()
c_neighbor.yfilter = YFilter.read

c_instance.neighbors.neighbor.append(c_neighbor)
isis.instances.instance.append(c_instance)
print(codec.encode(codec_provider, isis, subtree=True))
```

**Comment by thenetworkingguy** [comment]
Thank you for the response

The above snipper works fine.

With crud.read, I am using the top-level container (isis) to get the neighbor output. I was under the assumption that I need to put in the last key after reading a few posts on nested keys.

**Comment by abhikeshav** [comment]
No problem. With CRUD read, you can read any level of nested lists/containers with no leafs or keys specified. Just need to use YFilter.read. This has been implemented with ydk 0.6.0 and newer releases
